Lib/tty.py

```diff
@@ -15,22 +24,93 @@
 OSPEED = 5
 CC = 6
 
-def setraw(fd, when=TCSAFLUSH):
-    """Put terminal into a raw mode."""
-    mode = tcgetattr(fd)
-    mode[IFLAG] = mode[IFLAG] & ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
-    mode[OFLAG] = mode[OFLAG] & ~(OPOST)
-    mode[CFLAG] = mode[CFLAG] & ~(CSIZE | PARENB)
-    mode[CFLAG] = mode[CFLAG] | CS8
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON | IEXTEN | ISIG)
+def mkecho(mode, echo=True):
+    """Set/unset ECHO."""
+    if echo:
+        mode[LFLAG] |= ECHO
+    else:
+        mode[LFLAG] &= ~ECHO
+
+def mkraw(mode):
+    """Raw mode termios."""
+    mode[IFLAG] &= ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
+    mode[OFLAG] &= ~(OPOST)
+    mode[CFLAG] &= ~(CSIZE | PARENB)
+    mode[CFLAG] |= CS8
+    mode[LFLAG] &= ~(ECHO | ICANON | IEXTEN | ISIG)
     mode[CC][VMIN] = 1
     mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+
+def setraw(fd, when=TCSAFLUSH):
+    """Put terminal into raw mode.
+    Returns original termios."""
+    mode = tcgetattr(fd)
+    new = list(mode)
+    mkraw(new)
+    tcsetattr(fd, when, new)
+    return mode
 
 def setcbreak(fd, when=TCSAFLUSH):
-    """Put terminal into a cbreak mode."""
+    """Put terminal into cbreak mode.
+    Returns original termios."""
     mode = tcgetattr(fd)
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON)
-    mode[CC][VMIN] = 1
-    mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+    new = list(mode)
+    new[LFLAG] &= ~(ECHO | ICANON)
+    new[CC][VMIN] = 1
+    new[CC][VTIME] = 0
+    tcsetattr(fd, when, new)
+    return mode
+
+def login_tty(fd):
+    """Makes the calling process a session leader; if fd is a
+    descriptor of a tty, then that tty becomes the stdin, stdout,
+    stderr, and controlling tty of the calling process. Closes fd."""
+    # Establish a new session.
+    os.setsid()
+
+    # The tty becomes stdin/stdout/stderr.
+    os.dup2(fd, STDIN_FILENO)
+    os.dup2(fd, STDOUT_FILENO)
+    os.dup2(fd, STDERR_FILENO)
+    if fd > STDERR_FILENO:
+        os.close(fd)
+
+    # The tty becomes the controlling terminal.
+    try:
+        ioctl(STDOUT_FILENO, TIOCSCTTY)
+    except (NameError, OSError):
+        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
```

Comment:
I'm not sure what this line is doing exactly? Maybe add a comment here?

Reply:
Here is the source of `login_tty(3)`: https://sourceware.org/git/?p=glibc.git;a=blob;f=login/login_tty.c

The
```
    except (NameError, OSError):
        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
```
in our code is equivalent to their

```
 {
        /* This might work.  */
        char *fdname = ttyname (fd);
        int newfd;
        if (fdname)
          {
            if (fd != 0)
              (void) close (0);
            if (fd != 1)
              (void) close (1);
            if (fd != 2)
              (void) close (2);
            newfd = open (fdname, O_RDWR);
            (void) close (newfd);
          }
 }
```

I contacted the glibc people a while ago and this is what they had to say

> Since this was added by Roland almost 25 years ago, it seems to be trying to
emulate the TIOCSCTTY for system derived from System V (as indicated by this
stackflow thread [1]).

> Back in the day glibc tries to enclose fallbacks to be portable to the various
Unix implementations.  Currently this is just a dead code with various
shortcomings (as for a lot of the glibc fallback implementation where proper
kernel support was the correct answer) since neither Linux nor Hurd uses it.

> And it seems that Unixes derived from System V also has moved on from this
interface, neither Solaris nor Illumos seems to provide it.

> https://urldefense.com/v3/__https://stackoverflow.com/questions/51593530/code-explanation-for-glibc-login-tty-function-openttyname-immediately-f__;!!KwNVnqRv!Vtu1n8f09GnQ28wafOTRPvyTWxkgO0bd7c3sHMxR5GQqPjq4jJhaL_8J8kw-ZliJ$

TL;DR: TIOCSCTTY is available everywhere now. Therefore, I too think that it would be safe to remove the `except (NameError, OSError):` block. *Should I remove it?* I am going to use `tty.login_tty()` [ or whatever we end up naming it ] to replace the fallback code in `pty.fork()`, which still uses the old System V approach. `tty.login_tty()` will also be used in `pty.spawn()` to make it possible for us to implement some new/improved functionality.